File: langs/java-ddd/src/Navigator.java
final class Navigator {
    static void execute(ScentRepository scents, Grid grid, Robot robot, String instr) {
        for (int i = 0; i < instr.length() && !robot.lost; i++) {
            char c = instr.charAt(i);
            if (c == 'L')
                robot.turnLeft();
            else if (c == 'R')
                robot.turnRight();
            else if (c == 'F') {
                Position next = robot.nextForward();
                if (!grid.contains(next)) {
                    if (!scents.has(robot.position, robot.orientation)) {
                        scents.add(robot.position, robot.orientation);
                        robot.lost = true;
                    }
                } else {
                    robot.position = next;
                }
            }
        }
    }
}

File: langs/go-ddd/internal/domain/navigator.go
package domain

func Execute(repo ScentRepository, grid Grid, robot *Robot, instr string) {
	for _, c := range instr {
		if robot.Lost {
			break
		}
		switch c {
		case 'L':
			robot.TurnLeft()
		case 'R':
			robot.TurnRight()
		case 'F':
			next := robot.NextForward()
			if !grid.Contains(next) {
				if !repo.Has(robot.Position, robot.Orientation) {
					repo.Add(robot.Position, robot.Orientation)
					robot.Lost = true
				}
			} else {
				robot.Position = next
			}
		}
	}
}
